app/graph.go

```diff
@@ -0,0 +1,193 @@
+package app
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	acltypes "github.com/cosmos/cosmos-sdk/x/accesscontrol/types"
+)
+
+type DagNode struct {
+	NodeID          int
+	TxIndex         int
+	AccessOperation acltypes.AccessOperation
+}
+
+type DagEdge struct {
+	FromNodeID      int
+	ToNodeID        int
+	AccessOperation *acltypes.AccessOperation
+}
+
+type Dag struct {
+	NodeMap      map[int]DagNode
+	EdgesMap     map[int][]DagEdge                  // maps node Id (from node) and contains edge info
+	AccessOpsMap map[acltypes.AccessOperation][]int // tracks latest node to use a specific access op
+	TxIndexMap   map[int]int                        // tracks latest node ID for a tx index
+	NextID       int
+}
+
+func (edge *DagEdge) GetCompletionSignal() *CompletionSignal {
+	// only if there is an access operation
+	if edge.AccessOperation != nil {
+		return &CompletionSignal{
+			FromNodeID:      edge.FromNodeID,
+			ToNodeID:        edge.ToNodeID,
+			AccessOperation: *edge.AccessOperation,
+		}
+	}
+	return nil
+}
+
+// Order returns the number of vertices in a graph.
+func (dag Dag) Order() int {
+	return len(dag.NodeMap)
+}
+
+// Visit calls the do function for each neighbor w of vertex v, used by the graph acyclic validator
+func (dag Dag) Visit(v int, do func(w int, c int64) (skip bool)) (aborted bool) {
+	for _, edge := range dag.EdgesMap[v] {
+		// just have cost as zero because we only need for acyclic validation purposes
+		if do(edge.ToNodeID, 0) {
+			return true
+		}
+	}
+	return false
+}
+
+func NewDag() Dag {
+	return Dag{
+		NodeMap:      make(map[int]DagNode),
+		EdgesMap:     make(map[int][]DagEdge),
+		AccessOpsMap: make(map[acltypes.AccessOperation][]int),
+		TxIndexMap:   make(map[int]int),
+		NextID:       0,
+	}
+}
+
+func (dag *Dag) AddNode(txIndex int, accessOp acltypes.AccessOperation) DagNode {
+	dagNode := DagNode{
+		NodeID:          dag.NextID,
+		TxIndex:         txIndex,
+		AccessOperation: accessOp,
+	}
+	dag.NodeMap[dag.NextID] = dagNode
+	dag.NextID++
+	return dagNode
+}
+
+func (dag *Dag) AddEdge(fromIndex int, toIndex int, accessOp *acltypes.AccessOperation) *DagEdge {
+	// no-ops if the from or to node doesn't exist
+	if _, ok := dag.NodeMap[fromIndex]; !ok {
+		return nil
+	}
+	if _, ok := dag.NodeMap[toIndex]; !ok {
+		return nil
+	}
+	newEdge := DagEdge{fromIndex, toIndex, accessOp}
+	dag.EdgesMap[fromIndex] = append(dag.EdgesMap[fromIndex], newEdge)
+	return &newEdge
+}
+
+func (dag *Dag) AddNodeBuildDependency(ctx sdk.Context, txIndex int, accessOp acltypes.AccessOperation) {
```

Comment:
Since we're adding the edge to the previous accessOperation (either read or write), would this be an issue when that transaction is not the last operation in a given message 

(i.e don't we need to wait for write operations to commit in a message before reading it?)

Reply:
Gotcha, thanks for clarifying!